src/imitation/scripts/analyze.py

```diff
@@ -74,62 +123,170 @@ def gather_tb_directories(
     """Parse Sacred logs and generate a DataFrame for imitation learning results.
 
+    This function calls the helper `_gather_sacred_dicts`, which captures its arguments
+    automatically via Sacred. Provide those arguments to select which Sacred
+    results to parse.
+
     Args:
-      source_dir: A directory containing Sacred FileObserver subdirectories
-        associated with the `train_adversarial` Sacred script. Behavior is
-        undefined if there are Sacred subdirectories associated with other
-        scripts.
-      run_name: If provided, then only analyze results from Sacred directories
-        associated with this run name. `run_name` is compared against the
-        "experiment.name" key in `run.json`.
-      skip_failed_runs: If True, then filter out runs where the status is FAILED.
-      csv_output_path: If provided, then save a CSV output file to this path.
-      verbose: If True, then print the dataframe.
+        csv_output_path: If provided, then save a CSV output file to this path.
+        tex_output_path: If provided, then save a LaTeX-format table to this path.
+        print_table: If True, then print the dataframe to stdout.
+        table_verbosity: Increasing levels of verbosity, from 0 to 2, increase the
+            number of columns in the table.
 
     Returns:
-      A list of dictionaries used to generate the analysis DataFrame.
+        The DataFrame generated from the Sacred logs.
     """
-    sacred_dicts = _get_sacred_dicts(source_dir, run_name, env_name, skip_failed_runs)
+    table_entry_fns_subset = _get_table_entry_fns_subset(table_verbosity)
 
     rows = []
-    for sd in sacred_dicts:
+    for sd in _gather_sacred_dicts():
         row = OrderedDict()
         rows.append(row)
-
-        # Use get to prevent exceptions when reading in-progress experiments.
-        row["status"] = get(sd.run, "status")
-        row["use_gail"] = get(sd.config, "init_trainer_kwargs.use_gail")
-        row["env_name"] = get(sd.config, "env_name")
-        row["n_expert_demos"] = get(sd.config, "n_expert_demos")
-        row["run_name"] = get(sd.run, "experiment.name")
-
-        imit_stats = get(sd.run, "result.imit_stats")
-        expert_stats = get(sd.run, "result.expert_stats")
-        if imit_stats is not None and expert_stats is not None:
-            # Assume that `result.imit_stats` and `result.expert_stats` are
-            # formatted correctly.
-            row["expert_return_summary"] = _make_return_summary(expert_stats)
-            row["imit_return_summary"] = _make_return_summary(imit_stats, "monitor_")
-            row["imit_expert_ratio"] = (
-                imit_stats["monitor_return_mean"] / expert_stats["return_mean"]
-            )
+        for col_name, make_entry_fn in table_entry_fns_subset.items():
+            row[col_name] = make_entry_fn(sd)
 
     df = pd.DataFrame(rows)
+    if len(df) > 0:
+        assert set(df.columns) == set(table_verbosity_mapping[table_verbosity])
```

Comment:
This fails if `table_verbosity >= len(table_verbosity_mapping)`, in which case all of `table_entry_fns` are used.